datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -657,7 +662,13 @@ pub async fn from_substrait_rel(
                         table: nt.names[2].clone().into(),
                     },
                 };
-                let t = ctx.table(table_reference).await?;
+
+                let substrait_schema =
+                    from_substrait_named_struct(named_struct, extensions)?
+                        .replace_qualifier(table_reference.clone());
+
+                let t = ctx.table(table_reference.clone()).await?;
+                let t = ensure_schema_compatability(t, substrait_schema)?;
                 let t = t.into_optimized_plan()?;
                 extract_projection(t, &read.projection)
```

Comment:
I started wondering if the `ensure_schema_compatability` can now conflict with `extract_projection` - and I think it can, either by failing if DF doesn't optimize the select into a projection, or if DF does, then by overriding the select's projection with the Substrait projection...

I guess a fix would be something like in `extract_projection`, if there is an existing `scan.projection`, then apply `columnIndices` on it first

Reply:
That did indeed cause problems. It triggered an error of `unexpected plan for table` in `extract_projection`.

I added some code for this case in [`d571eb2` (#12245)](https://github.com/apache/datafusion/pull/12245/commits/d571eb2784ff54617483ebe4435bbff47dd6bea4). Is something like this what you had in mind?

I am noticing that the plans generated look a little weird/bad with a lot of redundant projects
```
                "Projection: DATA.a, DATA.b\
                \n  Projection: DATA.a, DATA.b\
                \n    Projection: DATA.a, DATA.b, DATA.c\
                \n      TableScan: DATA projection=[b, a, c]"
```
but they are at least correct for now.

Reply:
> Is something like this what you had in mind?

What I had in mind was manipulating the `scan.projection` directly - kinda like it is alreadydone in extract_projection, we could do it that way also for `ensure_schema_compatibility`. That way there wouldn't be additional Projections, and maybe it'd be a bit more efficient if the current setup doesn't push the column-pruning into the scan level (though I'm a bit surprised they don't get optimized anyways). 

But I don't think it's necessary - the way you've done it here seems correct, and we (I?) can do the project-mangling as a followup, unless you want to take a stab at it :)

Reply:
I think project unmangling would be better as a follow-up. Possible as part of https://github.com/apache/datafusion/issues/12347 because supporting remaps is going to add yet another layer of Projects 😅